### PR TITLE
React unit testing

### PR DESCRIPTION
**Pull request by AjroudiRida**
In this featured branch we did unit testing for the react components
-We used Jest snapshots to test the components.
-We used React Testing Library to simulate user interaction.
-We tested the calculator logic file

### CODE REVIEW

File: src/logic/__test__/calculate.test.js [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+    const obj = {
+      total: '6',
+      next: '2',
+      operation: '÷',
+    };
+    const btn = '=';
+    const result = calculate(obj, btn);
+    expect(result.total).toBe('3');
+  });
+});
```
Comment (Munsa1) [inline src/logic/__test__/calculate.test.js]: - [ ] Good job writing these unit tests. However I notice that you are importing the `calculate.js` to test these operations, but you should import the `operate.js` for these operations as it is responsible for the operations.
Once you do this you will notice that you have the following missing tests:

- A test for the modulo `%` operation
- A test that you cannot find the modulo divided by `0`
- A test for unknown operations (i.e you cant operate on non numerical data)
- A test for you cant divide a number by 0
-

File: src/logic/__test__/calculate.test.js [hunk 2]
```diff
@@ -0,0 +1,48 @@
+import calculate from '../calculate';
+
```
Comment (Munsa1) [inline src/logic/__test__/calculate.test.js]: - [ ] You will then need to write separate tests for the `calculate.js` using react testing library to simulate user interactions to test all the scenarios involved:

- If the buttonName is 'AC', the function resets the calculator data object to its initial state by setting total, next, and operation to null.
- If buttonName is a number (checked using the isNumber function), the function updates the calculator data object based on the current state and the button pressed. Depending on whether there is an ongoing operation or not, the next and total properties are updated to build the number to be operated on.
- If buttonName is '.', the function handles decimal input. It ensures that a decimal point is added to either the next or total property only if it's not already present.
- If buttonName is '=', the function performs the calculation based on the current state (if there's a pending operation) and updates the total property while clearing next and operation.
- If buttonName is '+/-', the function toggles the sign of the current number (either in next or total) by multiplying it by -1.
- If buttonName is any other operation (e.g., +, -, *, /), the function handles the logic for applying the operation. It considers cases like when an operation is pressed immediately after = or when there's already an ongoing operation.



Check out an example of how to write the tests:

```
import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import Calculator from './Calculator'; // Assuming you have a Calculator component

describe('Calculator', () => {
  it('resets calculator data when AC button is clicked', () => {
    const { getByText, getByTestId } = render(<Calculator />);
    
    const totalDisplay = getByTestId('total-display');
    const nextDisplay = getByTestId('next-display');
    const operationDisplay = getByTestId('operation-display');
    
    fireEvent.click(getByText('1'));
    fireEvent.click(getByText('+'));
    fireEvent.click(getByText('AC'));

    expect(totalDisplay).toHaveTextContent('Total: null');
    expect(nextDisplay).toHaveTextContent('Next: null');
    expect(operationDisplay).toHaveTextContent('Operation: null');
  });

  it('handles number input and operations correctly', () => {
    const { getByText, getByTestId } = render(<Calculator />);
    
    const totalDisplay = getByTestId('total-display');
    const nextDisplay = getByTestId('next-display');
    const operationDisplay = getByTestId('operation-display');
    
    fireEvent.click(getByText('1'));
    fireEvent.click(getByText('+'));
    fireEvent.click(getByText('2'));
    fireEvent.click(getByText('='));
    
    expect(totalDisplay).toHaveTextContent('Total: 3');
    expect(nextDisplay).toHaveTextContent('Next: null');
    expect(operationDisplay).toHaveTextContent('Operation: null');
  });

  it('handles decimal input correctly', () => {
    const { getByText, getByTestId } = render(<Calculator />);
    
    const nextDisplay = getByTestId('next-display');
    
    fireEvent.click(getByText('2'));
    fireEvent.click(getByText('.'));
    fireEvent.click(getByText('5'));
    
    expect(nextDisplay).toHaveTextContent('Next: 2.5');
  });

  it('handles +/- button correctly', () => {
    const { getByText, getByTestId } = render(<Calculator />);
    
    const nextDisplay = getByTestId('next-display');
    
    fireEvent.click(getByText('1'));
    fireEvent.click(getByText('+/-'));
    
    expect(nextDisplay).toHaveTextContent('Next: -1');
  });

  it('performs calculations correctly', () => {
    const { getByText, getByTestId } = render(<Calculator />);
    
    const totalDisplay = getByTestId('total-display');
    const nextDisplay = getByTestId('next-display');
    const operationDisplay = getByTestId('operation-display');
    
    fireEvent.click(getByText('3'));
    fireEvent.click(getByText('+'));
    fireEvent.click(getByText('2'));
    fireEvent.click(getByText('='));
    
    fireEvent.click(getByText('-'));
    fireEvent.click(getByText('1'));
    fireEvent.click(getByText('='));
    
    expect(totalDisplay).toHaveTextContent('Total: 4');
    expect(nextDisplay).toHaveTextContent('Next: null');
    expect(operationDisplay).toHaveTextContent('Operation: null');
  });

  // Add more test cases for different scenarios...

});
```

**Review by Munsa1 — CHANGES_REQUESTED** [review]
Hi @AjroudiRida @MozamelJawad , 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!
<FEEL FREE TO CHANGE THIS MESSAGE TO YOUR PERSONAL WORDS>

## Highlights
Passing Test Cases :heavy_check_mark: 
SnapShot Tests For All Components :heavy_check_mark: 
Descriptive Pr :heavy_check_mark: 
Good Communication :heavy_check_mark: 

## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

You can also consider:
- x
- y
- z

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Review by ridwanediallo — APPROVED** [review]
# STATUS: APPROVED 🟢
## Hi team, 
Congratulations!GIF

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

## (Highlights) Good Points: :+1:
- You documented your work professionally. ✔
- Well done implementing the requested changes. ✔
- The milestone's requirements have been met with success. 🎉 🎉 🎉 ✔

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**